include/arch/arm/arch/64/mode/machine.h

```diff
@@ -350,28 +350,61 @@ static inline word_t PURE getFAR(void)
 static inline word_t ats1e2r(word_t va)
 {
     word_t par;
+
+    UNUSED word_t saved_par;
+    if (config_set(CONFIG_ARM_HYPERVISOR_SUPPORT)) {
+        MRS("par_el1", saved_par);
+        isb();
+    }
+
     asm volatile("at s1e2r, %0" :: "r"(va));
     isb();
     MRS("par_el1", par);
+
+    if (config_set(CONFIG_ARM_HYPERVISOR_SUPPORT)) {
+        MSR("par_el1", saved_par);
+    }
     return par;
 }
 
 static inline word_t ats1e1r(word_t va)
 {
     word_t par;
+
+    UNUSED word_t saved_par;
+    if (config_set(CONFIG_ARM_HYPERVISOR_SUPPORT)) {
+        MRS("par_el1", saved_par);
+        isb();
+    }
+
     asm volatile("at s1e1r, %0" :: "r"(va));
     isb();
     MRS("par_el1", par);
+
+    if (config_set(CONFIG_ARM_HYPERVISOR_SUPPORT)) {
+        MSR("par_el1", saved_par);
+    }
```

Comment:
Could you add two inline functions that are defined to nothing when `!CONFIG_ARM_HYPERVISOR_SUPPORT` to avoid the code repetition? That way you can just add calls to those to all functions and make them nicer to read, e.g.

```
word_t saved_par = arm_hyp_save_par();
...
arm_hyp_restore_par(saved_par);
```

Reply:
@Indanz this is good suggestion.

Other alternative would be instead of adding it in the at function, we could add it in a higher level function. See (https://github.com/seL4/seL4/pull/1318/commits/a089a055531615d9b02d6427437cc1dfd7d98573)

also we need to make sure that future uses of this AT functions in other places in the kernel always save and restore PAR accordinly for the hypervisor mode.

Reply:
That's even better, that keeps those functions thin assembly wrappers.

I don't like how seL4 mixes external interface declarations/defines with internal stuff in header files. But basically no one outside of machine.h is using any of the ats* functions directly.